### PR TITLE
feat(HB): Progress for new exchange list

### PR DESCRIPTION
**Pull request by thisisalexmcgregor**
## Objective

This is essentially a progress update on the new `Exchange List` screen. It's not done but, I need to: 
1. Get feedback
2. Get some guidance on how we want to hook this into the `ExchangeCoordinator` etc. 

## Description

Most of this is just a Swift version of `ExchangeOrderViewController`. 

* `ExchangeTradeCellModel` is the model that we will be using for the cells that are displayed on this screen. It can be made with an `ExchangeTrade` or it can be decoded from a Homebrew endpoint. 
* `ExchangeListViewController` is the screen that will be replacing the `tableView` in `ExchangeOrderViewController`. It injects its `UITableView` into `ExchangeListDataProvider`. 
* `ExchangeListService` is the service that hits fetches the trades from the users' wallet. 
* `HomebrewExchangeService` is the service that hits the homebrew endpoint to fetch trades. It additionally supports pagination. 

## How to Test

At the moment you can't. Nothing is wired up just yet. There *will* be a PR though where you can test this and there *will* be a PR with unit tests. 

## Screenshot

trade_screenshot

## Related Information

* Ticket Number: IOS-1155

## Merge Checklist

- [x] The PR uses a title supported by [.changelogrc](https://github.com/blockchain/My-Wallet-V3-iOS/blob/dev/.changelogrc#L6...L69).
- [ ] Areas of technical debt are marked with a `// TICKET:` comment that includes a ticket number.
- [x] All unit tests pass.
- [ ] You have added unit tests.
- [x] New files added are within the correct directory. (e.g. if a file is required for unit tests to compile, be sure it is added to the tests target.)

### CODE REVIEW

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+             .expired,
+             .resolved:
+            return .red
+        }
+    }
+
+    var displayValue: String {
+        switch self {
+        case .complete:
+            return NSLocalizedString("Complete", comment: "").uppercased()
```
[the remark this answers is not in the document]
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: TIL that's a thing. Will do.

**Comment by thisisalexmcgregor** [comment]
@bchrisarriola Yes, in hindsight and as I was PR'ing this I realized that these could, and should be consolidated into one. How I'm going to do that I'm not sure just yet. Also, I'm not sure if `ExchangeCoordinator` should kick off these requests, wait until it receives a response, and based on that response present this screen **or** the `Exchange View` screen for either HB or SS.

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    enum TradeStatus: String {
+        case noDeposits = "no_deposits"
+        case received = "received"
+        case complete = "complete"
+        case resolved = "resolved"
+        case inProgress = "IN_PROGRESS"
+        case cancelled = "CANCELLED"
+        case failed = "failed"
+        case expired = "EXPIRED"
+        case none = "NONE"
```
[the remark this answers is not in the document]
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: We can, and I agree they are strange. They're like that because I was copying what's in `ExchangeTrade.h`. When I realized that the trades were coming from both the wallet **and** HB, that threw a wrench into things. Honestly that enum no longer needs to be a `String` type since we have the two init methods.

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift [hunk 3]
```diff
@@ -0,0 +1,178 @@
...
+//
+
+import Foundation
+
+struct ExchangeTradeCellModel: Decodable {
+
+    enum Currency: String, Decodable {
+        case btc = "BTC"
+        case eth = "ETH"
+        case bch = "BCH"
```
[the remark this answers is not in the document]
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: Oh. Didn't know that was a thing.

File: Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift [hunk 4]
```diff
@@ -0,0 +1,113 @@
...
+        let headerView = UINib(nibName: String(describing: ExchangeListOrderHeaderView.self), bundle: nil)
+        let listCell = UINib(nibName: ExchangeListViewCell.identifier, bundle: nil)
+        table.register(listCell, forCellReuseIdentifier: ExchangeListViewCell.identifier)
+        table.register(headerView, forHeaderFooterViewReuseIdentifier: String(describing: ExchangeListOrderHeaderView.self))
+    }
+
+    func display(page: Page<[ExchangeTradeCellModel]>) {
+        if let trades = page.result {
+            models = trades
+            tableView?.reloadData()
```
[the remark this answers is not in the document]
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift]: `Page` (which I don't like atm) does actually serve a purpose. If the `result` array on the `Page` has the same count as the `pageSize`, we don't request for additional trades as the user scrolls. It also has an optional `Error?`. It's not used at the moment but, I figured it may be useful. 

Also I wouldn't want to reload the whole table, I'd just want to reload the paths for the inserted rows. That's not in there yet either though.

**Comment by bchrisarriola** [comment]
> Also, I'm not sure if ExchangeCoordinator should kick off these requests, wait until it receives a response, and based on that response present this screen or the Exchange View screen for either HB or SS.

Seems like ExchangeCoordinator would be a good place to put that logic. On KYC we'll do something similar for the mid flow drop off

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift [hunk 5]
```diff
@@ -0,0 +1,178 @@
...
+    }
+
+    enum TradeStatus: String {
+        case noDeposits = "no_deposits"
+        case received = "received"
+        case complete = "complete"
+        case resolved = "resolved"
+        case inProgress = "IN_PROGRESS"
```
Comment (woudini) [inline Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: Looks like `TRADE_STATUS_IN_PROGRESS` can be removed. Doesn't seem to be used anywhere

File: Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+        return 1
+    }
+
+    func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
+        return models?.count ?? 0
+    }
+
+    func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
+        guard let model = models?[indexPath.row] else { return UITableViewCell() }
+        guard let cell = tableView.dequeueReusableCell(withIdentifier: ExchangeListViewCell.identifier, for: indexPath) as? ExchangeListViewCell else { return UITableViewCell() }
```
Comment (woudini) [inline Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift]: Swiftlint line length violation

File: Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+        cell.configure(with: model)
+        return cell
+    }
+}
+
+extension ExchangeListDataProvider: UITableViewDelegate {
+
+    func tableView(_ tableView: UITableView, viewForHeaderInSection section: Int) -> UIView? {
+        guard let header = tableView.dequeueReusableHeaderFooterView(withIdentifier: String(describing: ExchangeListOrderHeaderView.self)) as? ExchangeListOrderHeaderView else { return nil }
```
Comment (woudini) [inline Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift]: Swiftlint line length violation

File: Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift (hunk 6 above)
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/DataProviders/ExchangeListDataProvider.swift]: For whatever reason, Swiftlint doesn't show a warning on my end... Weird.

File: Blockchain/Homebrew/Exchange/Services/HomebrewExchangeService.swift [hunk 8]
```diff
@@ -0,0 +1,87 @@
...
+
+    func nextPage(fromTimestamp: Date, completion: @escaping (Page<[ExchangeTradeCellModel]>) -> ()) {
+        guard let baseURL = URL(string: BlockchainAPI.shared.retailCoreUrl) else { return }
+        let timestamp = DateFormatter.sessionDateFormat.string(from: fromTimestamp)
+        guard let endpoint = URL.endpoint(baseURL, pathComponents: ["trades"], queryParameters: ["before": timestamp]) else { return }
+        guard let session = NetworkManager.shared.session else { return }
+
+        var request = URLRequest(url: endpoint)
+        request.httpMethod = "GET"
+        request.allHTTPHeaderFields = ["Content-Type":"application/json",
```
Comment (woudini) [inline Blockchain/Homebrew/Exchange/Services/HomebrewExchangeService.swift]: Can use `HttpHeaderField.swift` here
```
            request.allHTTPHeaderFields = [
                HttpHeaderField.contentType: HttpHeaderValue.json,
                HttpHeaderField.accept: HttpHeaderValue.json
            ]
```

**Comment by woudini** [comment]
Should `AsyncOperation` and `AsyncBlockOperation` have an attributed license/copyright?

File: Blockchain/AssetType.swift [hunk 9]
```diff
@@ -25,6 +25,20 @@ extension AssetType {
...
+    init(stringValue: String) {
+        switch stringValue {
+        case "BTC":
+            self = .bitcoin
+        case "BCH":
+            self = .bitcoinCash
+        case "ETH":
+            self = .ethereum
+        default:
+            assertionFailure("\(stringValue) is not a supported type")
```
Comment (woudini) [inline Blockchain/AssetType.swift]: Would rather `fatalError` here and remove `self = .bitcoin`

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift (hunk 2 above)
[the remark this answers is not in the document]
Comment (woudini) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: >Honestly that enum no longer needs to be a String type since we have the two init methods.
Agree with this

File: Blockchain/Models/Wallet.m [hunk 10]
```diff
@@ -4071,6 +4072,7 @@ - (void)on_get_exchange_trades_success:(NSArray *)trades
         [exchangeTrades addObject:exchangeTrade];
     }
 
+    self.isFetchingExchangeTrades = NO;
```
Comment (woudini) [inline Blockchain/Models/Wallet.m]: `self.isFetchingExchangeTrades = NO;` should to be added to a new callback method `on_get_exchange_trades_error` that calls a `WalletExchangeDelegate` method (e.g. `didFailGetExchangeTrades`).
 Similarly to `objc_on_get_exchange_trades_success`, `objc_on_get_exchange_trades_error` should be called in `wallet-ios.js`. Ideally we would show an error alert with `[AlertViewPresenter.sharedInstance standardNotify...]`. Let me know if you have any questions on this https://github.com/blockchain/My-Wallet-V3-iOS/blob/f892e1503bcce7e532604717462d49617ed24d33/Blockchain/js/wallet-ios.js#L2550-L2553

TLDR: `isFetchingExchangeTrades` should be reset in error handling, just so happens there isn't any at the moment 😬

File: Blockchain/Homebrew/Exchange/Views/ExchangeListOrderHeaderView.swift [hunk 11]
```diff
@@ -0,0 +1,31 @@
...
+    fileprivate static let verticalPadding: CGFloat = 16.0
+
+    // MARK: Public
+
+    var actionHandler: (() -> Void)?
+
+    // MARK: Actions
+
+    @IBAction func newOrderTapped(_ sender: UIButton) {
+        if let block = actionHandler {
```
Comment (achtenhagen) [inline Blockchain/Homebrew/Exchange/Views/ExchangeListOrderHeaderView.swift]: Since `actionHandler` is declared optional, this `if` clause can be reduced to just `actionHandler?()`

File: Blockchain/AssetType.swift [hunk 12]
```diff
@@ -25,6 +25,20 @@ extension AssetType {
         }
     }
 
+    init(stringValue: String) {
+        switch stringValue {
+        case "BTC":
```
Comment (bchrisarriola) [inline Blockchain/AssetType.swift]: Might be better for this to fail a bit more gracefully by making this a failable initializer.

Also what about declaring a private static map so that the AssetType to symbol mapping is declared in one place (atm it's duplicated here and in the computed `symbol` property).

So:

```
enum AssetType: Int {
    private static let assetTypeToSymbolMap = [
        AssetType.bitcoin: "BTC",
        AssetType.bitcoinEth: "ETH",
        AssetType.bitcoinCash: "BCH",
    ]
}
```

...and use it in this initializer + also need to update `symbol`.

File: Blockchain/Homebrew/Exchange/API/ExchangeListAPI.swift [hunk 13]
```diff
@@ -0,0 +1,15 @@
+//
+//  ExchangeListAPI.swift
```
Comment (bchrisarriola) [inline Blockchain/Homebrew/Exchange/API/ExchangeListAPI.swift]: Update the file name + this?

File: Blockchain/Homebrew/Exchange/Services/ExchangeService.swift [hunk 14]
```diff
@@ -0,0 +1,150 @@
...
+    fileprivate let tradeQueue: OperationQueue = {
+        let queue = OperationQueue()
+        queue.qualityOfService = .userInitiated
+        return queue
+    }()
+    
+    override init() {
+        partnerAPI = PartnerExchangeService()
+        homebrewAPI = HomebrewExchangeService()
+        super.init()
```
Comment (bchrisarriola) [inline Blockchain/Homebrew/Exchange/Services/ExchangeService.swift]: Should be called first before init'ing the 2 services.

File: Blockchain/Homebrew/Exchange/Services/ExchangeService.swift [hunk 15]
```diff
@@ -0,0 +1,150 @@
...
+    }
+    
+    func getPartnerTrades() {
```
Comment (bchrisarriola) [inline Blockchain/Homebrew/Exchange/Services/ExchangeService.swift]: This should probably be `private` since we don't want to expose the implementation detail (same with `getHomebrewTrades`).

File: Blockchain/Homebrew/Exchange/Services/ExchangeService.swift [hunk 16]
```diff
@@ -0,0 +1,150 @@
...
+            models.forEach { [weak self] (model) in
+                guard let this = self else { return }
+                this.tradeModels.insert(model)
+            }
+        }
+    }
+    
+    fileprivate func sortAndUpdateTradeModels() {
+        let sorted = tradeModels.sorted(by: { $0.transactionDate.compare($1.transactionDate) == .orderedDescending })
+        let insertions: [ListPresentationUpdate<ExchangeTradeCellModel>] = sorted
```
Comment (bchrisarriola) [inline Blockchain/Homebrew/Exchange/Services/ExchangeService.swift]: So functional 😍

File: BlockchainTests/Operations/AsyncBlockOperationTests.swift [hunk 17]
```diff
@@ -0,0 +1,45 @@
...
+//  AsyncBlockOperationTests.swift
+//  BlockchainTests
+//
+//  Created by Alex McGregor on 8/22/18.
+//  Copyright © 2018 Blockchain Luxembourg S.A. All rights reserved.
+//
+
+import XCTest
+
+class AsyncBlockOperationTests: XCTestCase {
```
Comment (bchrisarriola) [inline BlockchainTests/Operations/AsyncBlockOperationTests.swift]: 💯

File: Blockchain/Homebrew/Exchange/Services/ExchangeService.swift (hunk 16 above)
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/Services/ExchangeService.swift]: `ListPresentationUpdate` works really nice when you want to abstract away a `NSFetchedResultsController`. Not that we would want to introduce Core Data 😬

**Comment by woudini** [comment]
@thisisalexmcgregor I see, so looks like we'd have to include Apple's license only if we didn't modify it.

**Comment by thisisalexmcgregor** [comment]
@woudini Right, definitely modified. There's [quite a few of these](https://github.com/dmcrodrigues/DRAsyncOperations).

File: Blockchain/Models/Wallet.m [hunk 18]
```diff
@@ -4080,6 +4084,15 @@ - (void)on_get_exchange_trades_success:(NSArray *)trades
     }
 }
 
+- (void)on_get_exchange_trades_error:(NSString *)error
+{
+    self.isFetchingExchangeTrades = NO;
+    if ([self.delegate respondsToSelector:@selector(didFailToGetExchangeTrades:)]) {
+        [self.delegate didFailToGetExchangeTrades:error];
+    }
+    [[AlertViewPresenter sharedInstance] standardNotifyWithMessage:error title:BC_STRING_ERROR in:nil handler:nil];
```
Comment (woudini) [inline Blockchain/Models/Wallet.m]: Would rather move to `partnerExchangeDidFailToGetTrades` and you can use `standardError` to automatically include an "Error" title (sorry forgot about that)

File: Blockchain/AssetType.swift (hunk 12 above)
Comment (woudini) [reply Blockchain/AssetType.swift]: +1 on using a static map, but not sure if we want the initializer to be failable, I'd rather `fatalError`

Comment (woudini) [reply Blockchain/AssetType.swift]: On the initializer failing, we can just display a generic error to try again later.

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift (hunk 5 above)
Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: I removed that `TRADE_STATUS_IN_PROGRESS` constant. Should I remove `inProgress` as well?

Comment (woudini) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: Yeah since the definition of "In progress" for Shapeshift is either `TRADE_STATUS_RECEIVED` or `TRADE_STATUS_NO_DEPOSITS`

File: Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift [hunk 19]
```diff
@@ -74,9 +74,13 @@ struct ExchangeTradeCellModel: Decodable {
 
     init(with trade: ExchangeTrade) {
         status = TradeStatus(shapeshift: trade.status)
-        assetType = AssetType(stringValue: trade.withdrawalCurrency())
         transactionDate = trade.date
         displayValue = trade.displayAmount() ?? ""
+        if let asset = AssetType(stringValue: trade.withdrawalCurrency()) {
+            assetType = asset
+        } else {
+            fatalError("Failed to map \(trade.withdrawalCurrency())")
```
Comment (woudini) [inline Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: Can `// TODO: show error alert` be added here?

Comment (thisisalexmcgregor) [reply Blockchain/Homebrew/Exchange/Models/ExchangeTradeCellModel.swift]: Sure, wasn't sure how we wanted to handle having a failable `AssetType`
